File: guardianwatch/lib/models/insight.dart
// ─── lib/models/insight.dart ─────────────────────────────────────────────────

enum InsightSeverity { normal, caution, warning, critical }

class Insight {
  final String id;
  final String title;
  final String summary;
  final String detail;
  final InsightSeverity severity;
  final bool isPremium;
  final DateTime generatedAt;
  final String? recommendation; // optional AI recommendation

  Insight({
    required this.id,
    required this.title,
    required this.summary,
    required this.detail,
    required this.severity,
    required this.isPremium,
    required this.generatedAt,
    this.recommendation,
  });

  Map<String, dynamic> toJson() => {
        'id': id,
        'title': title,
        'summary': summary,
        'detail': detail,
        'severity': severity.name,
        'is_premium': isPremium,
        'generated_at': generatedAt.toIso8601String(),
        if (recommendation != null) 'recommendation': recommendation,
      };

  factory Insight.fromJson(Map<String, dynamic> j) => Insight(
        id: j['id'] as String,
        title: j['title'] as String,
        summary: j['summary'] as String,
        detail: j['detail'] as String,
        severity: InsightSeverity.values.firstWhere(
          (e) => e.name == j['severity'],
          orElse: () => InsightSeverity.normal,
        ),
        isPremium: j['is_premium'] as bool? ?? false,
        generatedAt: DateTime.parse(j['generated_at'] as String),
        recommendation: j['recommendation'] as String?,
      );

  // For SQLite
  Map<String, dynamic> toMap() => {
        'id': id,
        'title': title,
        'summary': summary,
        'detail': detail,
        'severity': severity.name,
        'is_premium': isPremium ? 1 : 0,
        'generated_at': generatedAt.toIso8601String(),
        'recommendation': recommendation,
      };

  factory Insight.fromMap(Map<String, dynamic> m) => Insight(
        id: m['id'] as String,
        title: m['title'] as String,
        summary: m['summary'] as String,
        detail: m['detail'] as String,
        severity: InsightSeverity.values.firstWhere(
          (e) => e.name == m['severity'],
          orElse: () => InsightSeverity.normal,
        ),
        isPremium: (m['is_premium'] as int) == 1,
        generatedAt: DateTime.parse(m['generated_at'] as String),
        recommendation: m['recommendation'] as String?,
      );
}

File: guardianwatch-api/app/models/insight.py
"""
Insight model — mirrors `lib/models/insight.dart::Insight` field for field,
including the `severity` enum names (`normal|caution|warning|critical`)
which the Dart side parses with `InsightSeverity.values.firstWhere(...)`.
"""
import datetime as dt
import enum

from sqlalchemy import Boolean, Enum, ForeignKey, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.utils.db_types import UTCDateTime

from app.database import Base


class InsightSeverity(str, enum.Enum):
    normal = "normal"
    caution = "caution"
    warning = "warning"
    critical = "critical"


class Insight(Base):
    __tablename__ = "insights"

    id: Mapped[str] = mapped_column(String(64), primary_key=True)
    user_id: Mapped[str] = mapped_column(
        String(128), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )

    title: Mapped[str] = mapped_column(String(255), nullable=False)
    summary: Mapped[str] = mapped_column(Text, nullable=False)
    detail: Mapped[str] = mapped_column(Text, nullable=False)
    severity: Mapped[InsightSeverity] = mapped_column(
        Enum(InsightSeverity, native_enum=False, length=16),
        default=InsightSeverity.normal,
    )
    is_premium: Mapped[bool] = mapped_column(Boolean, default=False)
    recommendation: Mapped[str | None] = mapped_column(Text, nullable=True)

    generated_at: Mapped[dt.datetime] = mapped_column(
        UTCDateTime(), nullable=False, index=True
    )
    created_at: Mapped[dt.datetime] = mapped_column(
        UTCDateTime(), default=lambda: dt.datetime.now(dt.timezone.utc)
    )

    user: Mapped["User"] = relationship(back_populates="insights")  # noqa: F821
